homeassistant/const.py

```diff
@@ -411,8 +411,10 @@
 ENERGY_KILO_WATT_HOUR: Final[UnitEnergyT] = UnitEnergyT(UnitT("kWh"))
 
 # Electrical units
+ELECTRICAL_CURRENT_MILLIAMPERE: Final[UnitT] = UnitT("mA")
 ELECTRICAL_CURRENT_AMPERE: Final[UnitT] = UnitT("A")
 ELECTRICAL_VOLT_AMPERE: Final[UnitT] = UnitT("VA")
+ELECTRIC_POTENTIAL_MILLIVOLT: Final[UnitT] = UnitT("mV")
```

Comment:
Only comment I have, is `ELECTRIC` vs `ELECTRICAL`, I think we should pick a single one.

Reply:
I would like to suggest just `ELECTRIC`, but would be fine with either. Furthermore, I would like to reorganize these in a future PR.
* Rename `VOLT`
* `ELECTRICAL_VOLT_AMPERE` should really be a `POWER` unit as it's equivalent to `POWER_WATT`.
* Add new types: `ElectricCurrentT` and `ElectricPotentialT`

Reply:
I think we should lose the `T` postfix.

Reply:
Maybe we should even consider using ENUMs instead of defining all these types.

Reply:
> Maybe we should even consider using ENUMs instead of defining all these types.

I though about that while preparing the initial PR (#53147). The reason against it was that it's difficult to add any other values after the fact (not in `const.py`). In addition, using them just as strings would be more complicated. In the end, from a typing perspective, `NewType` seemed to be the best option.

Just an example, here the value could be modified to `UnitT("people")` and it would type check. (It currently works anyway as the entity base still uses `str`, this PR is one part of the update.)
https://github.com/home-assistant/core/blob/dd8ec04e5800aeed41ed0aeb52bce84811bf224d/homeassistant/components/coronavirus/sensor.py#L27-L30

> I think we should lose the T postfix.

Usually `T` is the naming convention for type aliases, and by extension `NewTypes`. For `Unit` it might be possible, but consider something like `DeviceClass`, `Event`, or `Service` where it would also make sense to add `NewType`.

--
The whole idea behind the change is to be able to better type Sensor metadata, like described in #53147. An excellent example for that is #53168 which already uses the new `UnitT` type.

Reply:
I don't see the need or use, what does it add? We don't restrict units at all and any string (or None) is a valid unit.

Using enum's can add additional value for the future (e.g., make conversion between units easier), or even introduce something like `pint` instead. Those are things that add value IMHO.

This only adds fluff to something that is simply a string.

Reply:
It's true there isn't a need for it per se. That does not however mean there is no use for it!

IMO it will add three things.
* First, catch errors where strings are assigned to `unit_of_measurement` attributes by mistake.
* Second, improve type hints in any form of mapping (dict / TypedDict, NamedTuple, Dataclass).
* And third, require developers to be more careful with units from a third party library as they can no longer be used and will require a mapping. This will make our code more resilient and most of all help prevent inconsistencies in the long run.

An example, this change is from the PR with which I added the Units in the first place: [homeassistant/components/lcn/climate.py -> diff](https://github.com/home-assistant/core/pull/53124/files#diff-e2abfa7f7db48d5d5ddcca7fd011962101dacdd931cb1340d288adfd223d05f3L110-R118)

Another from one of the first PRs to use `dataclasses` for sensor metadata: https://github.com/home-assistant/core/blob/40807d32006cd7b3602644e042be5cd48d9c73c9/homeassistant/components/climacell/const.py#L154-L161

Reply:
Also, any string is valid, we don't dictate its contents and we don't require it to be defined in our constants even. There is no way to enforce this and will only add fluff and more complexity that had no function.

I strongly disagree and don't like this change.

Reply:
> There is no way to enforce this in a consistent way

There is, I just haven't gotten around to changing the code as I need to fix other issue for it first (like this one). It's only necessary to modify the types in the `Entity` base class.

https://github.com/home-assistant/core/blob/450fdc91e47fe9cf76224db8141098e265ad0b79/homeassistant/helpers/entity.py#L238

https://github.com/home-assistant/core/blob/450fdc91e47fe9cf76224db8141098e265ad0b79/homeassistant/helpers/entity.py#L314-L315

--
There might be more places, but modifying these two alone will result in the following mypy errors
```
Return type "Optional[str]" of "unit_of_measurement" incompatible with return type "Optional[UnitT]" in supertype "Entity"  [override]
```
```
Incompatible types in assignment (expression has type "str", base class "Entity" defined the type as "Optional[UnitT]")  [assignment]
```

---
I understand your criticism and would like to propose having a discussion in the architecture repo to hear what others think of it. I should be able to write something up in the next days. If the decision there is to revert it, I'll do that. Please do keep in mind that I wouldn't have proposed the change if I didn't think it could add some value for us.

Reply:
>  if I didn't think it could add some value for us.

I realize that, however, we differ in opinion on the added value. I think the juice isn't worth the squeeze and inconsistency because of added complexity that doesn't encourage starting/first-timers to jump in.

So IMHO, instead of helping, this is becoming harmful.

Reply:
Additionally, if we really want to add value here, we should start using a system like `pint` and enforce units to actually be defined and convertible. There are a couple of architectural issues open for that as well.

IMHO, that would be adding value (well... value conversion 😉 )

Reply:
> I think the juice isn't worth the squeeze and inconsistency because of added complexity that doesn't encourage starting/first-timers to jump in.

These are still strings. So unless you add typing (which by itself is advanced) it only changes how they're defined. It would also only be enforced through typing. First-timers can continue to use strings.

--
I'll do the architecture discussion for it.

Reply:
>  First-timers can continue to use strings.

The problem is, is a lot of existing code is "borrowed" and copy-pasted together from all around the code base. We've seen this time and time again. Having inconsistent signatures show up in those cases and cause problems for these group of people.

Reply:
@frenck Though a night about it. Although I still think these new types provide a benefit, the concerns you mentioned are valid, too. I will open a PR to revert the change. Thank you for taking the time to discuss it!

Some of the arguments that convinced me:
* The added value is minimal
* Additional complexity for first-time contributors and devs not familiar with `NewType`
* Complicate any future decision to move to Enums / different unit management system
* A dozen or so integrations would required time consuming refactorings, which can introduce new bugs.